File: web/src/services/api/questions.ts
import { api } from './client'

export type QuestionCreate = {
  title: string
  description?: string
  category_id?: string
}

export type QuestionUpdate = {
  title?: string
  description?: string
  category_id?: string
  status?: string
}

export type QuestionResponse = {
  id: string
  user_id: string
  title: string
  description: string | null
  category_id: string | null
  status: string
  ai_analysis_status: string
  created_at: string
  updated_at: string | null
}

export type QuestionListResponse = {
  items: QuestionResponse[]
  total: number
  page: number
  page_size: number
}

export const questionsApi = {
  list: (params?: { page?: string; page_size?: string; category_id?: string; status?: string }) =>
    api.get<QuestionListResponse>('/api/v1/questions', params),

  getById: (id: string) =>
    api.get<QuestionResponse>(`/api/v1/questions/${id}`),

  create: (data: QuestionCreate) =>
    api.post<QuestionResponse>('/api/v1/questions', data),

  update: (id: string, data: QuestionUpdate) =>
    api.patch<QuestionResponse>(`/api/v1/questions/${id}`, data),

  delete: (id: string) =>
    api.delete<void>(`/api/v1/questions/${id}`),
}

File: backend/app/services/questions.py
from fastapi import HTTPException, status

from app.repositories.questions import QuestionRepository
from app.schemas.questions import QuestionCreate, QuestionUpdate, QuestionResponse, QuestionListResponse


class QuestionService:
    def __init__(self, repo: QuestionRepository):
        self.repo = repo

    async def create(self, user_id: str, data: QuestionCreate) -> QuestionResponse:
        question = await self.repo.create(
            user_id=user_id,
            title=data.title,
            description=data.description,
            category_id=data.category_id,
        )
        return QuestionResponse.model_validate(question)

    async def get_by_id(self, question_id: str) -> QuestionResponse:
        question = await self.repo.get_by_id(question_id)
        if not question:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Question not found")
        return QuestionResponse.model_validate(question)

    async def list(
        self,
        page: int = 1,
        page_size: int = 20,
        category_id: str | None = None,
        status: str | None = None,
    ) -> QuestionListResponse:
        items, total = await self.repo.list(
            page=page,
            page_size=page_size,
            category_id=category_id,
            status=status,
        )
        return QuestionListResponse(
            items=[QuestionResponse.model_validate(q) for q in items],
            total=total,
            page=page,
            page_size=page_size,
        )

    async def update(self, question_id: str, data: QuestionUpdate, user_id: str) -> QuestionResponse:
        question = await self.repo.get_by_id(question_id)
        if not question:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Question not found")
        if str(question.user_id) != user_id:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Not authorized to update this question")
        update_data = {k: v for k, v in data.model_dump(exclude_none=True).items()}
        updated = await self.repo.update(question_id, update_data)
        return QuestionResponse.model_validate(updated)

    async def delete(self, question_id: str, user_id: str) -> None:
        question = await self.repo.get_by_id(question_id)
        if not question:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Question not found")
        if str(question.user_id) != user_id:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Not authorized to delete this question")
        await self.repo.soft_delete(question_id)
